### DISCUSSION TITLE
Opening file in xxd or any other program

### DISCUSSION THREAD

**Discussion by cool-RR** [saulpw/visidata#1842] [Q&A] [answered]
Today I opened a file in VisiData, after browsing through a directory structure to find it. The file was binary, so VisiData's text-showing capabilities were not very useful for me.

I wanted to open the file in a hex viewer such as xxd. **What would be an efficient way to do that?** I could quit VisiData and run xxd in the shell, but then I'd have to find the file all over again, and I'll lose my VisiData session.

In general, I think we should have the ability to say "Open this file in the given program", regardless of whether it's xxd or any other program.

**Comment by saulpw** [comment, marked as answer]
```
@VisiData.api
def open_external(vd, path, prog):
    import subprocess
    with SuspendCurses():
        subprocess.call(['sh', '-i', '-c', f'{prog} {path}; exec sh'])

BaseSheet.addCommand('', 'open-external', 'open_external(source, input("open with: "))')
```

10 minutes

**Reply by cool-RR** [reply]
That works, thanks. The annoying bit is that `xxd` won't page. After some trial and error, I found that a good way is to type `Ctrl-O` to open with Vim, and then use `:%!xxd` to open `xxd` within Vim and get the paging. 

One thing that bothers me is that `Ctrl-O` (i.e. `sysopen-row`) can only work from the folder view, and not when you already opened the file, which is how your `open_external` works. But I can live with that.

**Reply by saulpw** [reply]
`BaseSheet.addCommand('', 'sysopen-source', 'launchEditor(source)')`

To explain this piece by piece:

- `addCommand` is how we add new commands to a VisiData Sheet
- putting a command on `BaseSheet` means that it is available to all sheets (since all sheets, including non-tabular sheets, derive from BaseSheet).  In develop (slated for v3.0) there will be `vd.addCommand` as an alias for `BaseSheet.addCommand`, to indicate that they are global commands (`globalCommand` will still work, but that's been deprecated)
- The first string is for the keybinding; an empty string `''` means that it is unbound, and so you can only execute it by the longname
- The longname is `sysopen-source`; `sysopen` is the verb, which means to open using a system program (external to VisiData), and `source` is the noun
- Every sheet has a `source`, which is usually a Path (for the original data sheets) or the parent sheet (for derived sheets).
- The command string (or "execstr") is Python code, which can access both `vd` and `sheet` attributes without needing to specify the `vd` or the `sheet`.  With the explicit objects, this code would be `vd.launchEditor(sheet.source)`.
- `launchEditor` is used in the DirSheet `sysopen-row` as you noticed; we can look at the execstr for that command by pressing `z Ctrl+H` on the DirSheet to see all commands, pressing `gv` on that sheet to unhide the execstr column, and then finding the `sysopen-row` command.
- The type of each row (the "rowdef") on the DirSheet is a visidata.Path; this is how `launchEditor(cursorRow)` works for sysopen-row.
- Since `launchEditor` just takes a Path, and the `source` for some newly loaded data sheet is a Path, we can call `launchEditor(source)` and it will just work.
- For a complete command, there would be a separate help string after the execstr, indicating what the command does in English.

and that's it!

**Reply by daviewales** [reply]
This only works when `source` is a Path.
Would it make sense / be possible to recursively search back through sources until the source is a Path?
(Might be confusing for joined tables, but should work OK for filtered ones?)

**Reply by saulpw** [reply]
Yes, maybe we could make a `rootSource` property that works as you say.  I'm a bit wary, as it is too tempting to make a `save-source` command that saves the current sheet to the rootSource, which would be disastrous if you happened to be on a filtered sheet.  But maybe that is just paranoia :)  If you want to make a PR for rootSource (you could model it after `rootSheet()`)  and to add an updated `sysopen-source`, I think we could merge it.
